mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,785 @@
+        }
+    }};
+}
+
+pub trait FactoryTrait<F>: Fn() -> F {}
+
+impl<Factory, F> FactoryTrait<F> for Factory where Factory: Fn() -> F {}
+
+pub trait RepeatableTrait<T, E>: Future<Output = Action<T, E>> + Unpin {}
+
+impl<F, T, E> RepeatableTrait<T, E> for F where F: Future<Output = Action<T, E>> + Unpin {}
+
+pub(crate) trait InspectErrorTrait<E>: 'static + Fn(&E) + Send {}
+
+impl<F: 'static + Fn(&E) + Send, E> InspectErrorTrait<E> for F {}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum RepeatError<E> {
+    TimeoutExpired {
+        until: Until,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+    AttemptsExceed {
+        attempts: usize,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+}
+
+impl<E> RepeatError<E> {
+    pub fn error(&self) -> &E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    pub fn into_error(self) -> E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    fn timeout(until: Until, error: E) -> Self { RepeatError::TimeoutExpired { until, error } }
+
+    fn attempts(attempts: usize, error: E) -> Self { RepeatError::AttemptsExceed { attempts, error } }
+}
+
+impl<E: fmt::Display> fmt::Display for RepeatError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RepeatError::TimeoutExpired { until, error } => {
+                write!(
+                    f,
+                    "Waited too long until {until:?} for the future to succeed. Error: {error}",
+                )
+            },
+            RepeatError::AttemptsExceed { attempts, error } => {
+                write!(f, "Error {error} on retrying the future after {attempts} attempts")
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Until {
+    Instant(Instant),
+    TimestampMs(u64),
+}
+
+/// The future is ether ready (with a `T` result), or not ready (failed with an intermediate `E` error).
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
+    factory: Factory,
+    /// Currently executable future, i.e. an active attempt.
+    exec_fut: Option<F>,
```

Comment:
Is there a specific reason for it to be `Option`? Everything seems fine with the following diff

```diff
index ba7e428bb..f2728dbce 100644
--- a/mm2src/common/custom_futures/repeatable.rs
+++ b/mm2src/common/custom_futures/repeatable.rs
@@ -188,7 +188,7 @@ impl<T, E> RetryOnError<T, E> for Result<T, E> {
 pub struct Repeatable<Factory, F, T, E> {
     factory: Factory,
     /// Currently executable future, i.e. an active attempt.
-    exec_fut: Option<F>,
+    exec_fut: F,
     /// A timeout future if we're currently waiting for a timeout.
     timeout_fut: Option<Timer>,
     until: Option<Until>,
@@ -209,7 +209,7 @@ where
 
         Repeatable {
             factory,
-            exec_fut: Some(exec),
+            exec_fut: exec,
             timeout_fut: None,
             until: None,
             attempts: None,
@@ -329,8 +329,7 @@ where
                 return Poll::Pending;
             }
 
-            let mut exec = self.exec_fut.take().unwrap();
-            match exec.poll_unpin(cx) {
+            match self.exec_fut.poll_unpin(cx) {
                 Poll::Ready(Action::Ready(ready)) => return Poll::Ready(Ok(ready)),
                 Poll::Ready(Action::Retry(error)) => {
                     if let Some(ref inspect) = self.inspect_err {
@@ -359,7 +358,7 @@ where
                     }
 
                     // Create a new future attempt.
-                    self.exec_fut = Some((self.factory)());
+                    self.exec_fut = (self.factory)();
                     // Set the timeout future if `self.repeat_every` is specified,
                     // otherwise the `self.exec_fut` will be repeated right at the next iteration.
                     self.timeout_fut = self.repeat_every.map(|timeout| Timer::sleep(timeout.as_secs_f64()));
@@ -367,7 +366,6 @@ where
                 },
                 // We should proceed with this `exec` future attempt.
                 Poll::Pending => {
-                    self.exec_fut = Some(exec);
                     return Poll::Pending;
                 },
             }
```

Reply:
Thanks! Will apply these changes!

Reply:
Done